Add tutorial project 22-23'

kfu02 opened the pull request:
## Description
New SW tutorial. Currently intended to go with general SW training--RC members will  be expected to watch the video lectures, for their general understanding of robotics topics, ROS, and C++, but to do this tutorial instead of completing the general coding SW tutorial.

Updated docs can be viewed here (open in incognito to avoid caching issues):
https://rj-rc-software.readthedocs.io/en/tutorial-project/

kfu02 committed:
add intro to tutorial, modify getting_started to account for it

kfu02 committed:
small tweaks

kfu02 committed:
add part 2. GitHub

kfu02 committed:
add ros basics

kfu02 committed:
finish section 5, concl, and stub out the rest

kfu02 committed:
Merge branch 'ros2' into tutorial-project

kfu02 committed:
clean up existing sections

kfu02 committed:
add gameplay section

kfu02 committed:
stub out last section, waiting on build instr

kfu02 committed:
add some pics

p-nayak11 committed:
Merge branch 'ros2' into tutorial-project

p-nayak11 committed:
Added the last section and fixed some outdated docstrings

p-nayak11 committed:
fixed hyperlinks and typo

p-nayak11 committed:
added more about CMake fiels, lambda exps & arrow pointers, and launch files

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -336,33 +335,96 @@ You can find the team color by subscribing to the relevant topic (this should
+
+The arrow operator is used to point the value on the right side of the arrow, usually an element of a structure or class, and assign it to the variable pointer name on the left side of arrow. Again, I HIGHLY recommend that you google and visit the following resources to learn more about this topic.
+
+    - `<https://www.journaldev.com/44327/arrow-operator-c-plus-plus>`_
```
honestly I think these links suck and this one explains it better: https://www.tutorialspoint.com/cplusplus/cpp_member_operators.htm#:~:text=The%20(%2D%3E)%20arrow%20operator&text=The%20%2D%3E%20is%20called%20the%20arrow,pointer%2C%20use%20the%20arrow%20operator.

kfu02 replied:
see above suggestion, I disagree with your description of ->

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -336,33 +335,96 @@ You can find the team color by subscribing to the relevant topic (this should
+
+   robot_status_pubs_.at(robot_id)->publish(robot_status);
+
+The arrow operator is used to point the value on the right side of the arrow, usually an element of a structure or class, and assign it to the variable pointer name on the left side of arrow. Again, I HIGHLY recommend that you google and visit the following resources to learn more about this topic.
```
```suggestion
The arrow operator is used to access a method or element of an object, when given a pointer to that object. Above, ``robot_status_pubs_`` is a list of pointers to ROS publisher objects. Calling `->publish(robot_status)` on one element in that list publishes a robot status using that specific publisher. 

You will learn about pointers when you take CS 2110. For now, you can use the typical dot operator (e.g. ``robot_status_pubs_.at(0).publish(...)``) and the compiler will tell you when you should use the arrow operator instead. If you would like to learn more now, see `this source <https://www.tutorialspoint.com/cplusplus/cpp_member_operators.htm#:~:text=The%20(%2D%3E)%20arrow%20operator&text=The%20%2D%3E%20is%20called%20the%20arrow,pointer%2C%20use%20the%20arrow%20operator.>`_.
```

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -336,33 +335,96 @@ You can find the team color by subscribing to the relevant topic (this should
+    - `<https://www.journaldev.com/44327/arrow-operator-c-plus-plus>`_
+    - `<https://www.geeksforgeeks.org/arrow-operator-in-c-c-with-examples/>`_
+
+* Notice the docstrings for the radio header file. It explains that it's the abstract superclass of the network_radio and sim_radio node. If you are unfamiliar with the concept of abstraction, `here <https://www.pythontutorial.net/python-oop/python-abstract-class/>` is more information. When exploring these two nodes, you will discover a third file named "node"_main.cpp which contains the main fu ...
```
```suggestion
* Notice the docstrings for the radio header file. It explains that it's the abstract superclass of the network_radio and sim_radio node. If you are unfamiliar with the concept of abstraction, `here <https://www.pythontutorial.net/python-oop/python-abstract-class/>`_ is more information. When exploring these two nodes, you will discover a third file named "node"_main.cpp which contains the main function for its respective node. This structure is intended to make writing the CMakeLists.txt file for the directory easier.
```
link not formatted right

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -336,33 +335,96 @@ You can find the team color by subscribing to the relevant topic (this should
 
+Similar to the python section, there's a lot of file-finding in this part. Use the option in your IDE or text editor that allows you to see a full folder at once. For instance, in VS Code, there is an option to open a full folder, which
+displays all the subfolders and files in the left toolbar. If you open
+``robocup-software/rj_gameplay`` like this, it should be a lot easier to
```
rj_gameplay?

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -336,33 +335,96 @@ You can find the team color by subscribing to the relevant topic (this should
-files.
+files. For more information, check `this <https://cplusplus.com/articles/Gw6AC542/>`_ resource.
 
 That may have confused you. Let's take a look at a real example in our codebase
```
```suggestion
Let's take a look at a real example in our codebase
```
I know I wrote this line but I think it's wordy on a second glance

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -336,33 +335,96 @@ You can find the team color by subscribing to the relevant topic (this should
-// explain which parts of the .hpp are actually relevant to setting up the node
-// explain where they should put stuff into the build system (do this yourself
-to make sure it's right)
+* There are multiple `#include`s in both files. Reference between the ROS tutorial above and your objective to determine which files are needed for your node.
```
```suggestion
* There are multiple ``#include``s in both files. Think of these as equivalent to `import` in Python or Java. Reference between the ROS tutorial above and your objective to determine which files are needed for your node.
```

kfu02 replied:
@p-nayak11 honestly I don't find it necessary for the new members to figure this out on their own, how do you feel about just giving them all the right includes, namespace and maybe an empty class? I worry that it may be frustrating to learn C++ syntax

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -336,33 +335,96 @@ You can find the team color by subscribing to the relevant topic (this should
+                manipulators_cached_.at(i) = *manipulator;
+            });
+
+A `lambda expression` is used in place of the topic callback function in the ROS tutorial. The lambda expression allows you to pass in-line function objects and requires less lines of code when compared to having another function. I HIGHLY recommend that you google and visit the following resources to learn more about this topic.
```
```suggestion
A ``lambda expression`` is used in place of the topic callback function in the ROS tutorial. The lambda expression allows you to pass in-line function objects and requires less lines of code when compared to having another function. I HIGHLY recommend that you google and visit the following resources to learn more about this topic.
```

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -336,33 +335,96 @@ You can find the team color by subscribing to the relevant topic (this should
+
+**Key Sections**
+
+* Notice all the source files in the ``ROBOCUP_LIB_SRC``. You will find the radio files that you explored earlier along with other crucial source files that aid in motion control, soccer ui, etc. You will get familiar with these files as you continue to contribute to the codebase.
```
again I think it may be best to lower the difficulty by giving them exact lines to put into CMake files, we're trying to get them to interface with ROS pub/sub and maybe pick up on some C++ syntax, not learn our whole build system (which I don't even really know)

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -336,33 +335,96 @@ You can find the team color by subscribing to the relevant topic (this should
+
+You may have started noticing the pattern of adding to the key sections of the CMake files. You need to find the parts of the file where the other nodes are located and follow the format for your own node. While it is important to understand the meaning of each section, you will eventually learn as you gain experience. As explained earlier, CMake files are build files for the directory. So check  ...
+
+You're almost there! One more step to go before acquiring all the necessities to create your own node. Launch files in ROS are a convenient way of starting up multiple nodes, setting initial parameters, and other requirements. Find the launch directory and explore the file that would pertain to where your node would be located. As you already know, follow the pattern for the other nodes that are  ...
```
```suggestion
You're almost there! One final file you have to create is a launch file. Launch files in ROS are a convenient way of starting up multiple nodes, setting initial parameters, and other requirements. Find the launch directory and explore the file that would pertain to where your node would be located. As you already know, follow the pattern for the other nodes that are functionally similar to your node.
```
which launch directory? specific file path would make this easier

kfu02 replied:
and or specific nodes to look at launch files for

p-nayak11 replied:
I will add where the launch files are located and guide them to the particular file. However, telling them which node to copy the format for is unnecessary because most of the nodes follow the same format. Common sense would tell them to follow what most of the nodes do.

kfu02 reviewed (commented):
@p-nayak11 some thoughts. ~~I think we should make this section a bit easier on them. I think I would've quit RoboCup if I was expected to learn C++, CMake, and ROS all at once. Maybe giving them an empty .cpp and .hpp file to copy-paste that just has all the boiler plate code done already--so hpp includes the right ROS stuff, .cpp includes the hpp, constructor is there already and pulls from rclcpp::Node, but the rest of it is up to them.~~

EDIT: upon further discussion in Slack, we have agreed that the difficulty is fine and we can always give hints/boilerplate if needed later. You can ignore the difficulty-related comments below.

Also, do you mind wrapping the lines of the .rst source files to 80 chars? That way it's a little easier to edit in the future. This is a VS code [extension](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap)that does it for you (or `gq<motion>` in Vim).

p-nayak11 committed:
rewrapped and made edits based on review

p-nayak11 committed:
missed one line to wrap

p-nayak11 commented:
> @p-nayak11 some thoughts. ~I think we should make this section a bit easier on them. I think I would've quit RoboCup if I was expected to learn C++, CMake, and ROS all at once. Maybe giving them an empty .cpp and .hpp file to copy-paste that just has all the boiler plate code done already--so hpp includes the right ROS stuff, .cpp includes the hpp, constructor is there already and pulls from rclcpp::Node, but the rest of it is up to them.~
> 
> EDIT: upon further discussion in Slack, we have agreed that the difficulty is fine and we can always give hints/boilerplate if needed later. You can ignore the difficulty-related comments below.
> 
> Also, do you mind wrapping the lines of the .rst source files to 80 chars? That way it's a little easier to edit in the future. This is a VS code [extension](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap)that does it for you (or `gq<motion>` in Vim).

This is so weird... I thought that all of the stuff that I added are automatically wrapped because of the settings that I have. For some of the lists, I guess it didn't work.

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -0,0 +1,511 @@
+
+**Key Sections**
+
+* There are multiple `#include`s in both files. Reference between the ROS
```
```suggestion
* There are multiple ``#include``s in both files. Reference between the ROS
```

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -0,0 +1,511 @@
+
+The arrow operator is used to access a method or element of an object, when
+given a pointer to that object. Above, ``robot_status_pubs_`` is a list of
+pointers to ROS publisher objects. Calling `->publish(robot_status)` on one
```
```suggestion
pointers to ROS publisher objects. Calling ``->publish(robot_status)`` on one
```

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -0,0 +1,511 @@
+pointers to ROS publisher objects. Calling `->publish(robot_status)` on one
+element in that list publishes a robot status using that specific publisher.
+
+    - `<https://www.tutorialspoint.com/cplusplus/cpp_member_operators.htm#:~:text=The%20(%2D%3E)%20arrow%20operator&text=The%20%2D%3E%20is%20called%20the%20arrow,pointer%2C%20use%20the%20arrow%20operator>`
```
this link is really ugly, can you make it inline (like "see [this source](url)")

p-nayak11 replied:
hyperlink broke :(, but yea i'll fix it

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -0,0 +1,511 @@
+  unfamiliar with the concept of abstraction, `here
+  <https://www.pythontutorial.net/python-oop/python-abstract-class/>`_ is more
+  information. When exploring these two nodes, you will discover a third file
+  named "node"_main.cpp which contains the main function for its respective
```
```suggestion
  named <node>_main.cpp which contains the main function for its respective
```
this is the standard way to show that the text given is not literal, just a filler

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -0,0 +1,511 @@
+  radio files that you explored earlier along with other crucial source files
+  that aid in motion control, soccer ui, etc. You will get familiar with these
+  files as you continue to contribute to the codebase.
+* Many of the nodes have an environment variable set for their "node"_main.cpp.
```
```suggestion
* Many of the nodes have an environment variable set for their <node>_main.cpp.
```

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -0,0 +1,511 @@
+
+You may have started noticing the pattern of adding to the key sections of the
+CMake files. You need to find the parts of the file where the other nodes are
+located and follow the format for your own node. While it is important to
```
```suggestion
located and follow the format for your own node. For now, it's okay if you don't understand everything that's going on.
```
sorry this probably ruins the wrapping

kfu02 commented on docs/source/tutorial.rst:
```diff
@@ -0,0 +1,511 @@
+You may have started noticing the pattern of adding to the key sections of the
+CMake files. You need to find the parts of the file where the other nodes are
+located and follow the format for your own node. While it is important to
+understand the meaning of each section, you will eventually learn as you gain
```
```suggestion
You will eventually learn as you gain
```

kfu02 reviewed (commented):
@p-nayak11 some minor suggestions. Looking great overall!

p-nayak11 committed:
changes requested

p-nayak11 committed:
Merge branch 'ros2' into tutorial-project

p-nayak11 committed:
hyperlink fix

p-nayak11 committed:
last one i missed

p-nayak11 reviewed (approved):
LGTM

fourpenny commented:
The tutorial overall looks great. Here's a couple things I'd suggest adding to hopefully reduce the number of repetitive/simple questions we get from new members:

**A link to our page on how to source the stack (the software repo README) or a hint for if they get an error related to sourcing** 
That is something that's easy to forget to do (and even easier to fix), but it'll likely prevent existing members from being swarmed by new people asking what's wrong when they don't source ROS (like when using the CLI tools in section 4) because it doesn't *explicitly* say to do so anywhere. I know the ROS tutorials mention this, but members may leave off in the middle of something or forget what resources are the best to check to remember how to do so, and if we leave a helpful link, it would probably be good.

**Reminders to test/run the sim when members write their play in section 3**
It seems a bit like a "duh" thing they should know how to do, but they may not think about it and feel slightly hesitant if they don't remember the exact process to launch it. This would hopefully get them used to the usual workflow when working on gameplay/UI stuff (type things, rebuild if applicable, launch sim and check the robot behavior) more quickly. I feel like this would be a second useful place to remind them of the sourcing process.

**Additional headers in section 6 (C++ and ROS)**
It's a long section, and people will likely take multiple meetings to get through both the node and CMakeLists.txt writing process. Not sure if the formatting style we are using for ReadTheDocs allows it, but I think it would help break up the long wall of (useful) text.

I'm happy to add anything above, so let me know if you think these are good ideas!

kasohrab reviewed (commented):
It looks great so far, a lot better than new member projects in my opinion.

I don't have much to add on to what Christian said, but I would suggest making a change for all the places where you all provide external guides and say "read through those sites before proceeding if you are unfamiliar."

I think we should leave those sources and encourage them to read them yes, but not suggest that they do that first. They should attempt each piece, then if they get stuck, they hopefully have become motivated or interested enough to want to go through these external sites. We don't want to demotivate potentially great members just because they got bored with external documentation and felt aimless. They should decide what they need to learn and why they need to learn it.

So the way I imagine the format is:

1. The challenges and motivation for this section
2. Implementation details
3. advice and hints for when you get stuck

It doesn't have to be quite like this, I just think this is generally a more fun way to structure a tutorial for someone following along as a new member.

Also if we could attach a video of our robots moving on real hardware (recent manual, recent autonomous motion, or that 2014 video lol) I think that would give more perspective to new software members on where their role could be.

p-nayak11 committed:
added changed requested

p-nayak11 commented:
> The tutorial overall looks great. Here's a couple things I'd suggest adding to hopefully reduce the number of repetitive/simple questions we get from new members:
> 
> **A link to our page on how to source the stack (the software repo README) or a hint for if they get an error related to sourcing** That is something that's easy to forget to do (and even easier to fix), but it'll likely prevent existing members from being swarmed by new people asking what's wrong when they don't source ROS (like when using the CLI tools in section 4) because it doesn't _explicitly_ say to do so anywhere. I know the ROS tutorials mention this, but members may leave off in the middle of something or forget what resources are the best to check to remember how to do so, and if we leave a helpful link, it would probably be good.
> 
> **Reminders to test/run the sim when members write their play in section 3** It seems a bit like a "duh" thing they should know how to do, but they may not think about it and feel slightly hesitant if they don't remember the exact process to launch it. This would hopefully get them used to the usual workflow when working on gameplay/UI stuff (type things, rebuild if applicable, launch sim and check the robot behavior) more quickly. I feel like this would be a second useful place to remind them of the sourcing process.
> 
> **Additional headers in section 6 (C++ and ROS)** It's a long section, and people will likely take multiple meetings to get through both the node and CMakeLists.txt writing process. Not sure if the formatting style we are using for ReadTheDocs allows it, but I think it would help break up the long wall of (useful) text.
> 
> I'm happy to add anything above, so let me know if you think these are good ideas!

The Getting Started page addressed issues with sourcing and installing our stack. I believe the starting of the tutorial also points them towards it. I also included the FAQ page in the list of options for help with errors and debugging. That solves some common issues with sourcing and whatnot. I added the second and third suggestions in the last commit.

p-nayak11 committed:
Merge branch 'ros2' into tutorial-project

kfu02 committed:
clean up C++ section, tighten voice, add some missing backticks

kfu02 commented:
@kasohrab sadly it's not so easy to put video on a static website. Rest assured we will show them fun videos/live robots on day 1.

kfu02 committed:
fix broken links, tweak Cmake intro

kfu02 commented:
hey @p-nayak11 , I changed some of your section to make the flow a bit better from previous sections and enforce my internal style (e.g. no bare links, active voice, more references to the concrete SimRadio class). lmk if you disagree with something I changed

and @fourpenny I broke the C++ part into even smaller subsections now, so hopefully it's more readable. agreed that it used to be too long

kfu02 commented:
merging now for GI
